src/meta/load_balance_policy.h

```diff
@@ -0,0 +1,254 @@
+                if (primary_count > replicas_high) {
+                    higher_count++;
+                } else if (primary_count < replicas_low) {
+                    lower_count++;
+                }
+            }
+
+            if (0 == higher_count && 0 == lower_count) {
+                return nullptr;
+            }
+            return dsn::make_unique<ford_fulkerson>(
+                _app, _nodes, _address_id, higher_count, lower_count, replicas_low);
+        }
+
+    private:
+        const std::shared_ptr<app_state> &_app;
+        const node_mapper &_nodes;
+        const std::unordered_map<dsn::rpc_address, int> &_address_id;
+    };
+
+private:
+    void make_graph();
+    void add_edge(int node_id, const node_state &ns);
+    void update_decree(int node_id, const node_state &ns);
+    void handle_corner_case();
+
+    int select_node(std::vector<bool> &visit, const std::vector<int> &flow);
+    int max_value_pos(const std::vector<bool> &visit, const std::vector<int> &flow);
+    void update_flow(int pos,
+                     const std::vector<bool> &visit,
+                     const std::vector<std::vector<int>> &network,
+                     std::vector<int> &flow,
+                     std::vector<int> &prev);
+
+    const std::shared_ptr<app_state> &_app;
+    const node_mapper &_nodes;
+    const std::unordered_map<dsn::rpc_address, int> &_address_id;
+    uint32_t _higher_count;
+    uint32_t _lower_count;
+    int _replicas_low;
+    size_t _graph_nodes;
+    std::vector<std::vector<int>> _network;
+
+    FRIEND_TEST(ford_fulkerson, add_edge);
+    FRIEND_TEST(ford_fulkerson, update_decree);
+    FRIEND_TEST(ford_fulkerson, find_shortest_path);
+    FRIEND_TEST(ford_fulkerson, max_value_pos);
+    FRIEND_TEST(ford_fulkerson, select_node);
+};
+
+class copy_replica_operation
+{
+public:
+    copy_replica_operation(const std::shared_ptr<app_state> app,
+                           const app_mapper &apps,
+                           node_mapper &nodes,
+                           const std::vector<dsn::rpc_address> &address_vec,
+                           const std::unordered_map<dsn::rpc_address, int> &address_id);
+    virtual ~copy_replica_operation() = default;
+
+    bool start(migration_list *result);
+
+protected:
+    void init_ordered_address_ids();
+    virtual int get_partition_count(const node_state &ns) const = 0;
+
+    gpid select_partition(migration_list *result);
+    const partition_set *get_all_partitions();
+    gpid select_max_load_gpid(const partition_set *partitions, migration_list *result);
+    void copy_once(gpid selected_pid, migration_list *result);
+    void update_ordered_address_ids();
+    virtual bool only_copy_primary() = 0;
+    virtual bool can_select(gpid pid, migration_list *result) = 0;
+    virtual bool can_continue() = 0;
+    virtual balance_type get_balance_type() = 0;
+
+    std::set<int, std::function<bool(int a, int b)>> _ordered_address_ids;
```

Comment:
better to give `a` and `b` a meaningful name

Reply:
It's like the i in `for (int i = 0; i < 100; i++)`, so it's hard to give a meaningful name

Reply:
```suggestion
    // ref: https://github.com/XiaoMi/rdsn/pull/953/files#diff-2aa231a206c212f7198f9cfa4ae5a42731fd01a6d5a793983ada6454543db891R716
    std::set<int, std::function<bool(int partition_index_l, int partition_index_r)>> _ordered_address_ids;
```